### ISSUE TITLE
Time-domain map was not aligned

### ISSUE THREAD

**Issue by scnu2631** [zhaodongsun/contrast-phys#7]
Hello, after generating the .npy file using test.py, I tried to visualize it and plot the time-domain graph. However, I found that the predicted rPPG signal does not align with the GT (Ground Truth) signal. Below are the images and code. Do you have any suggestions?
The code：
```
import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import welch, resample
from utils_sig import hr_fft, butter_bandpass, normalize

# 读取 .npy 文件
data = np.load(r'C:\first-year graduate\contrast-phys-master\results\2\3\subject38_pred.npy', allow_pickle=True)
rppg_list = data.item().get('rppg_list')
bvp_list = data.item().get('bvp_list')

# 展平信号
rppg = np.concatenate(rppg_list)
bvp = np.concatenate(bvp_list)

# 标准化信号
rppg_norm = normalize(rppg)
bvp_norm = normalize(bvp)

# 计算信号长度和采样率
signal_length = len(rppg_norm)
fs = 30  # 假设采样率为 30Hz

# 对标准化信号进行滤波
rppg_filtered = butter_bandpass(rppg_norm, lowcut=0.6, highcut=4, fs=fs)
bvp_filtered = butter_bandpass(bvp_norm, lowcut=0.6, highcut=4, fs=fs)

# 计算预测心率和地面真值心率
hr_rppg, rppg_psd, rppg_hr = hr_fft(rppg_filtered, fs=fs)
hr_bvp, bvp_psd, bvp_hr = hr_fft(bvp_filtered, fs=fs)

# 计算评估指标
mae = np.mean(np.abs(rppg_filtered - bvp_filtered))
mse = np.mean((rppg_filtered - bvp_filtered) ** 2)
rmse = np.sqrt(mse)
r = np.corrcoef(rppg_filtered, bvp_filtered)[0, 1]

# 绘制时域图和频谱图
fig, (ax1, ax2) = plt.subplots(2, figsize=(20, 10))

# 时域图
time = np.arange(signal_length) / fs
ax1.plot(time, rppg_filtered, label='Predicted Signal')
ax1.plot(time, bvp_filtered, label='Ground Truth Signal')
ax1.set_xlabel('Time (s)')
ax1.set_ylabel('Amplitude')
ax1.set_title('Predicted Signal vs Ground Truth Signal')
ax1.legend()
ax1.grid('on')

# 频谱图
ax2.plot(rppg_hr, rppg_psd, label='Predicted Signal')
ax2.plot(bvp_hr, bvp_psd, label='Ground Truth Signal')
ax2.set_xlabel('Heart Rate (bpm)')
ax2.set_ylabel('Power Spectral Density')
ax2.set_xlim([40, 200])
ax2.set_title('Power Spectral Density of Predicted Signal vs Ground Truth Signal')
ax2.legend()
ax2.grid('on')

# 显示评估指标和预测心率
print(f"MAE: {mae:.4f}")
print(f"MSE: {mse:.4f}")
print(f"RMSE: {rmse:.4f}")
print(f"Pearson's Correlation Coefficient (r): {r:.4f}")
print(f"Predicted Heart Rate: {hr_rppg:.2f} bpm")
print(f"Ground Truth Heart Rate: {hr_bvp:.2f} bpm")

plt.show()
```
![Figure_1](https://github.com/zhaodongsun/contrast-phys/assets/97783238/27a3fb0d-9efd-49ad-b82b-679e35e7e885)

**Comment by zhaodongsun** [comment]
Hi, 

this is a good question. During training, the signals are tranformed to the frequency domains, and the loss function is performend on the the PSDs of the signals. However, when the phase of the signal is changed (e.g., reverse phase as shown in your figure), the PSD won't change since the phase information is removed in the PSD  (only amplitude).

For supervised methods like PhysNet, they use Pearson correlation to compare GT with the output signal to align them in the temporal domain. Therefore, some supervised methods can achieve accurate alignment between GT and the output signal in the temporal domain. However, during rPPG data collection, GT should be accurated synchronized with the facial videos.

Hope this explaination can be helpful. Please feel free to request other clarification.

**Comment by scnu2631** [comment]
Thank you very much for your detailed explanation of the time-domain alignment issue. I understand that your method involves transforming the rPPG signal into the frequency domain and calculating the loss function on the Power Spectral Density (PSD) in that domain. The Power Spectral Density only contains the amplitude information of the signal, without the phase information. This indeed explains why there is a displacement between the final predicted result and the ground truth in the time-domain plot.

However, I would like to suggest something. Would it be possible to incorporate some auxiliary supervised signals during the training process? For example, using a small portion of well-labeled data to constrain the alignment of rPPG with BVP (Blood Volume Pulse) in the time domain? Or perhaps exploring some unsupervised time-domain alignment techniques? This might potentially enhance the accuracy of the predicted signal in the time domain while maintaining the unsupervised nature of your method.

Of course, these are just some of my thoughts. If they are not feasible or could lead to other issues, please feel free to correct me. Once again, thank you for your explanation, it has certainly deepened my understanding of the issue.

**Comment by zhaodongsun** [comment]
Hi,

These solutions might be possible ways. There is one simple solution. After training, you can compare one output signal with the GT to check whether the phase is reversed or not. In this way, you only need one GT signal to "calibrate" the phase.

**Comment by scnu2631** [comment]
Hi,
I believe this is a very clever and practical solution. Through this post-hoc calibration approach, the issue of alignment in the time domain for unsupervised methods can be effectively addressed without the need for significant modifications to the training process.

Certainly, there might be some finer details to consider in the implementation, such as determining phase reversals and efficiently performing batch calibration. However, overall, this does seem like a very promising approach worth exploring. Once again, thank you for suggesting this, it has given me a new perspective to consider.

**Comment by scnu2631** [comment]
Hello!

Thank you very much for your suggestion! I have added a phase calibration step in the test.py file, with the main modifications as follows:

1.Added a phase_calibration function to compare the phase of the rPPG signal with the corresponding BVP signal for each clip. With this modification, if the phase difference between rPPG and BVP is close to 0, the rPPG signal remains unchanged; if the phase difference is close to π, the rPPG signal is inverted.
```
def phase_calibration(rppg, bvp):
    # 计算rppg与bvp的相位差
    rppg_fft = np.fft.fft(rppg)
    bvp_fft = np.fft.fft(bvp)
    phase_diff = np.angle(rppg_fft) - np.angle(bvp_fft)

    # 检查相位差的均值是否接近π,如果是,则需要对rppg进行相位校准
    if np.abs(np.mean(phase_diff)) < 0.5 * np.pi:
        rppg_calibrated = -rppg
    else:
        rppg_calibrated = rppg

    return rppg_calibrated
```
2.After the model inference, I apply the phase_calibration function to the obtained rppg_clip for phase calibration, and store the calibrated rppg_clip into rppg_list.
 ```
for b in range(num_blocks):
                rppg_clip = dl_model(imgs[b*time_interval*fs:(b+1)*time_interval*fs])

                # 对当前clip进行相位校准
                bvp_clip = bvp[b * time_interval * fs:(b + 1) * time_interval * fs]
                rppg_clip = phase_calibration(rppg_clip, bvp_clip)

                rppg_list.append(rppg_clip)
                bvp_list.append(bvp[b*time_interval*fs:(b+1)*time_interval*fs])
```
As a result, I now have aligned time-domain plots!
![Figure_2](https://github.com/zhaodongsun/contrast-phys/assets/97783238/6c8ecdf5-853f-4821-a95c-3a2392ef4680)

Thank you again for your help, I have learned a lot from you！Wish you a happy life and smooth work!
